### PR TITLE
fix(parser): Parse more than utf8

### PR DESCRIPTION
**Pull request by sishtiaq**
fix(parse): Extend parser to parse more than UTF-8 (#954).

`sqlite3.c` has a char table that needs to be parsed. This PR allows non-UTF-8 to be parsed too. The three chars {\n,\t,\"} are still printed as human-readable, even if their corresponding hex codes are parsed. 

The main change is that `Parser.parseStringLiteral` now returns a `ByteArray`, rather than a `String`. Some of the previous uses (`AttrParser`, `MlirParser`)  converted to `Byte` anyway; these calculations have now gone. 
On the other hand, `UnitTest/Parser` has a wrapper to convert to `String` for comparison.

### CODE REVIEW

File: UnitTest/Parser.lean [hunk 1]
```diff
@@ -325,31 +331,40 @@ since single-byte values with a letter first digit (>= 0xA0) aren't valid UTF-8
...
+#eval testParser "\"hello world!\"" parseOptionalStringLiteralStr
+
+/-!
+{\t, \n, \"} and {\09, \0A, \22} are printed as {\t, \n, \"}.
+-/
+/--
+  info: "Success: (some (\t\n\"hello world\t\n\"))"
+-/
+#guard_msgs in
+#eval testParser "\"\\t\\n\\\"hello world\\09\\0A\\22\"" parseOptionalStringLiteralStr
```
Comment (math-fehr) [inline UnitTest/Parser.lean:368]: I guess you should add a test here where we return `none` from `parseOptionalStringLiteralStr`?

File: Veir/IR/Attribute.lean [hunk 2]
```diff
@@ -754,18 +754,37 @@ instance : ToString RegisterType where
...
-def escapeStringLiteral (s : String) : String := Id.run do
+-- def escapeStringLiteral (s : String) : String := Id.run do
+--   let mut result := ""
+--   for c in s.toList do
+--     if c == '\\' then result := result ++ "\\\\"
+--     else if c == '"' then result := result ++ "\\\""
+--     else if c == '\n' then result := result ++ "\\n"
+--     else if c == '\t' then result := result ++ "\\t"
+--     else result := result.push c
+--   return result
```
Comment (math-fehr) [inline Veir/IR/Attribute.lean]:
```suggestion
```

File: Veir/IR/Attribute.lean [hunk 3]
```diff
@@ -754,18 +754,37 @@ instance : ToString RegisterType where
...
+
+private def hexDigit (n : UInt8) : Char :=
+  if n < 10 then Char.ofNat (n.toNat + '0'.toNat)
+  else Char.ofNat (n.toNat - 10 + 'A'.toNat)
+
+def escapeStringLiteralBytes (b : ByteArray) : String := Id.run do
```
Comment (math-fehr) [inline Veir/IR/Attribute.lean]:
```suggestion
def escapeStringLiteral (b : ByteArray) : String := Id.run do
```
I think `Bytes` is not needed since we have it as argument already?

File: Veir/Parser/MlirParser.lean [hunk 4]
```diff
@@ -499,23 +499,24 @@ partial def parseOptionalOp (ip : Option InsertPoint) : MlirParserM (Option Oper
   let results ← parseOpResults
   let opNameStart ← getPos
   let some opName ← parseOptionalStringLiteral | return none
+  let opNameStr := String.fromUTF8! opName
```
Comment (math-fehr) [inline Veir/Parser/MlirParser.lean]: Here this will fail if we give non-utf8 operation name.
This was before failing with an "internal error" message.
Can you use `fromUTF8?` here and fail with an error message as well?

File: Veir/IR/Attribute.lean [hunk 5]
```diff
@@ -754,18 +754,37 @@ instance : ToString RegisterType where
 instance : ToString RegisterAttr where
   toString attr := s!"{attr.value} : !riscv.reg"
 
-def escapeStringLiteral (s : String) : String := Id.run do
+-- def escapeStringLiteral (s : String) : String := Id.run do
```
Comment (GZGavinZhao) [inline Veir/IR/Attribute.lean]: Stray code?

Comment (GZGavinZhao) [reply Veir/IR/Attribute.lean]: Nevermind, didn't read the PR description.

File: Veir/Parser/Parser.lean [hunk 6]
```diff
@@ -283,28 +283,27 @@ private def processEscapes (input : ByteArray) (basePos : Nat) : Except ParserEr
...
     let processed ← ofExcept (processEscapes raw (token.slice.start.byteOffset + 1))
-    match String.fromUTF8? processed with
-    | some str => return some str
-    | none => throwAt token.slice.start "internal error: failed converting string literal"
+    return some processed
   | none => return none
 
 /--
   Parse a string literal.
   Raise an error if the next token is not a string literal.
```
Comment (math-fehr) [inline Veir/Parser/Parser.lean]: You should also update this documentation AFAIK

**Review by math-fehr — COMMENTED** [review]
Thanks a lot! I added a few comments, but otherwise all good.

File: Veir/IR/Attribute.lean (hunk 5 above)
Comment (sishtiaq) [reply Veir/IR/Attribute.lean]: Removed commented-out code.

File: UnitTest/Parser.lean (hunk 1 above)
Comment (sishtiaq) [reply UnitTest/Parser.lean:368]: You mean an expected negative test like `testParser "123"` `=` `none`? If so, done.

File: Veir/IR/Attribute.lean (hunk 2 above)
Comment (sishtiaq) [reply Veir/IR/Attribute.lean]: Yep, gone now.

File: Veir/IR/Attribute.lean (hunk 3 above)
Comment (sishtiaq) [reply Veir/IR/Attribute.lean]: Renamed.

File: Veir/Parser/MlirParser.lean (hunk 4 above)
Comment (sishtiaq) [reply Veir/Parser/MlirParser.lean]: Done. I've included `opName` in the exn message, though I'm not sure if lean can toString it?

File: Veir/Parser/Parser.lean (hunk 6 above)
Comment (sishtiaq) [reply Veir/Parser/Parser.lean]: Do you mean to say that the comment should say "byte" rather than "string"? If so, done.

**Comment by sishtiaq** [comment]
I've addressed all comments @math-fehr @GZGavinZhao. If you can have a check, and put the PR in to the merge queue please.

**Review by math-fehr — APPROVED** [review]
Thanks, that looks good to me, I'll just wait for @GZGavinZhao to review it then I can merge it!

File: Veir/Parser/MlirParser.lean [hunk 7]
```diff
@@ -499,23 +499,25 @@ partial def parseOptionalOp (ip : Option InsertPoint) : MlirParserM (Option Oper
   let results ← parseOpResults
   let opNameStart ← getPos
   let some opName ← parseOptionalStringLiteral | return none
+  let some opNameStr := String.fromUTF8? opName
+    | throwAt opNameStart s!"op '{opName}' not a valid UTF8 string."
```
Comment (GZGavinZhao) [inline Veir/Parser/MlirParser.lean]: This would cause an invalid UTF8 error message to print the offending name as a decimal array: `op '[192, 128, 255]' not a valid UTF8 string.`. I think it would be slightly better UX if we do:
```
| throwAt opNameStart s!"op '{escapeStringLiteral opName}' is not a valid UTF-8 string."
```
so the error message looks like: `op '\C0\80\FF' is not a valid UTF-8 string.`

**Review by GZGavinZhao — APPROVED** [review]
Looks great! Thanks for implementing this. Just a small error message improvement suggestion, then should be good to merge!

File: Veir/Parser/MlirParser.lean (hunk 7 above)
Comment (sishtiaq) [reply Veir/Parser/MlirParser.lean]: Good point. Should use my own pretty-printer to, er, pretty-print.

Comment (sishtiaq) [reply Veir/Parser/MlirParser.lean]: Now done.

**Comment by math-fehr** [comment]
Thanks!
